### PR TITLE
Tests : devnet : declare and deploy account

### PR DESCRIPTION
**Pull request by ocdbytes**
Adding tests to devnet for declare and deploy transactions.

### CODE REVIEW

**Review by cchudant — APPROVED** [review]
very cool! nothing to say this is perfect :)

**Comment by cchudant** [comment]
note: this needs to be rebased onto main before merging

File: cairo/target/dev/madara_contracts_madara_contracts_test_account_AccountUpgradeable.contract_class.json
Comment (cchudant) [inline cairo/target/dev/madara_contracts_madara_contracts_test_account_AccountUpgradeable.contract_class.json:1]: this should not be staged in, see https://github.com/madara-alliance/madara/pull/238#discussion_r1736623501

**Review by cchudant — CHANGES_REQUESTED** [review]

File: crates/client/devnet/src/lib.rs
Comment (cchudant) [inline crates/client/devnet/src/lib.rs:1]: we may be able to remove the trust_class_hashes flag then?

**Review by cchudant — APPROVED** [review]
fixed the cairo artifacts problem, it was my fault (:
lgtm :partying_face::partying_face::partying_face:

File: crates/client/devnet/src/lib.rs [hunk 1]
```diff
@@ -325,6 +335,162 @@ mod tests {
...
+                paymaster_data: vec![],
+                account_deployment_data: vec![],
+                nonce_data_availability_mode: DataAvailabilityMode::L1,
+                fee_data_availability_mode: DataAvailabilityMode::L1,
+                is_query: false,
+            });
+
+        let res = chain.sign_and_add_declare_tx(declare_txn, sender_address);
+
+        let calculated_class_hash = sierra_class.clone().class_hash().unwrap();
```
Comment (apoorvsadana) [inline crates/client/devnet/src/lib.rs]: can we avoid clone here?

File: crates/client/devnet/src/lib.rs [hunk 2]
```diff
@@ -325,6 +335,162 @@ mod tests {
...
+        assert_eq!(res.class_hash, calculated_class_hash);
+
+        chain.block_production.set_current_pending_tick(1);
+        chain.block_production.on_pending_time_tick().unwrap();
+
+        let block = chain.backend.get_block(&BlockId::Tag(BlockTag::Pending)).unwrap().unwrap();
+
+        assert_eq!(block.inner.transactions.len(), 1);
+        assert_eq!(block.inner.receipts.len(), 1);
+        log::info!("receipt: {:?}", block.inner.receipts[0]);
```
Comment (apoorvsadana) [inline crates/client/devnet/src/lib.rs]: debug?

File: crates/client/devnet/src/lib.rs [hunk 3]
```diff
@@ -325,6 +335,162 @@ mod tests {
...
+
+        let class_info =
+            chain.backend.get_class_info(&BlockId::Tag(BlockTag::Pending), &calculated_class_hash).unwrap().unwrap();
+        match class_info {
```
Comment (apoorvsadana) [inline crates/client/devnet/src/lib.rs]: can use `assert_matches!` directly

File: crates/client/devnet/src/lib.rs [hunk 4]
```diff
@@ -325,6 +335,162 @@ mod tests {
...
+            ClassInfo::Sierra(info) => {
+                assert_eq!(info.compiled_class_hash, compiled_contract_class_hash);
+            }
+            ClassInfo::Legacy(_) => {}
+        }
+
+        let TransactionReceipt::Declare(receipt) = block.inner.receipts[0].clone() else { unreachable!() };
+
+        assert_eq!(
```
Comment (apoorvsadana) [inline crates/client/devnet/src/lib.rs]: same as above

File: crates/client/devnet/src/lib.rs [hunk 5]
```diff
@@ -325,6 +335,162 @@ mod tests {
...
+            }
+        );
+    }
+
+    #[rstest]
+    fn test_account_deploy(mut chain: DevnetForTesting) {
+        println!("{}", chain.contracts);
+
+        let key = SigningKey::from_random();
+        log::info!("Secret Key : {:?}", key.secret_scalar());
```
Comment (apoorvsadana) [inline crates/client/devnet/src/lib.rs]: debug logs?

File: crates/client/devnet/src/lib.rs [hunk 6]
```diff
@@ -325,6 +335,162 @@ mod tests {
...
+            tip: 0,
+            paymaster_data: vec![],
+            nonce_data_availability_mode: DataAvailabilityMode::L1,
+            fee_data_availability_mode: DataAvailabilityMode::L1,
+            is_query: false,
+        });
+
+        let res = chain.sign_and_add_deploy_account_tx(deploy_account_txn, &account);
+
+        assert_eq!(res.contract_address, account.address);
```
Comment (apoorvsadana) [inline crates/client/devnet/src/lib.rs:478]: we should also check for txn receipt and see if it was succesful?

File: crates/client/devnet/src/lib.rs [hunk 7]
```diff
@@ -325,6 +335,162 @@ mod tests {
...
+                transaction_hash: res.transaction_hash,
+                messages_sent: vec![],
+                events: vec![Event {
+                    from_address: ERC20_STRK_CONTRACT_ADDRESS,
+                    keys: receipt.events[0].keys.clone(),
+                    data: receipt.events[0].data.clone(),
+                }],
+                execution_resources: receipt.execution_resources.clone(),
+                actual_fee: FeePayment { amount: receipt.actual_fee.amount, unit: PriceUnit::Fri },
+                execution_result: receipt.execution_result.clone(),
```
Comment (apoorvsadana) [inline crates/client/devnet/src/lib.rs]: with assert_matches! you can skip fields you don't want to check afaik. the `execution_result` should be hardcoded to check for success

File: crates/client/devnet/src/lib.rs (hunk 4 above)
Comment (ocdbytes) [reply crates/client/devnet/src/lib.rs]: Here we don't need assert matches as we are comparing the receipt.

**Review by apoorvsadana — APPROVED** [review]

**Pull request merged**